core/src/main/java/bio/terra/pearl/core/service/EnvironmentService.java

```diff
@@ -16,4 +17,14 @@ public EnvironmentService(EnvironmentDao dao) {
     public Optional<Environment> findOneByName(EnvironmentName name) {
         return dao.findByName(name);
     }
+
+    @Transactional
+    public Environment buildPersisted(Environment.EnvironmentBuilder builder, String testName) {
```

Comment:
I like this solution a lot! But I am not sure how I feel about these living outside of the tests. Hmm. Is it possible to make to make (or does it already exist?) a `TestDao` or something where Dao calls that are exclusively for tests can live? That way we only have to have the special extra SQL stuff when we need it.

Reply:
@connorlbark To help me understand better, are you more concerned about the changes in the `EnvironmentDao` and `getCreateQuerSql`, or about the way `buildPersisted` is being called instead of `create`?
If it's the former, I thought changing the sql query for this specific dao to do nothing in times of conflicts on env `name` is something that makes sense in other parts of the code as well.

Reply:
I see what you're saying about the fact that it shouldn't do anything in other cases, but I would argue that it is correct behavior for it to error if the application attempted to make a new environment with the same name, simply because it shouldn't happen and I would say it's a bug/unintentional call if it does. But I'm open to changing that view of things. I am mostly worried that, since the tests didn't work, we're changing application-level code, even though (imo) that code was totally fine and was doing the intended behavior. In addition, `buildPersisted` is a testing concept, and only exists within the tests - if we decide we want to not error on conflicts, which I am cool with, then I would just replace that call with a call to just the `create` function and remove `buildPersisted` for consistency.

Reply:
Ah now I see what you mean... That makes sense.